### PR TITLE
Reconciliation of MutatingWebhookConfiguration values didnt happen

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
Fix webhook spec update recognition

Enforce all the values we expect to be set on the webhooks. This ensures the hash
will always represent the full resource spec as we view it, so any specified field update
will be recognized and processed by the operator.

### CODE REVIEW

**Comment by Danil-Grigorev** [comment]
@JoelSpeed @enxebre This PR implements one of the possible approaches to handling webhooks. Merging this will allow dropping [generation management](https://github.com/openshift/machine-api-operator/blob/ab96046e110a7250caccf9ef7071b7935810931e/pkg/operator/webhook_configuration_helper.go#L86-L128) and spec hashing which is not currently used with our implementation, and rely solely on spec hash.

Another approach could be use generation comparison, and drop defaulting/spec hashing instead, but this way the operator will frequently poll API server for updates without value changes.

**Comment by JoelSpeed** [comment]
> Another approach could be use generation comparison, and drop defaulting/spec hashing instead, but this way the operator will frequently poll API server for updates without value changes.

Does the operator use an informer? If such, this shouldn't be a problem, we can use a cached reader and that should be sufficient.

If we have informers set up already for this, then we should consider if the generation approach may be simpler, I believe this is what other implementation in library go have been using as one of their indicators, there must be a good reason for that. We may want to speak to the API folks to try and work out a bit more about the history of the apply code

**Review by JoelSpeed — COMMENTED** [review]
Would be good to see some comments on the defaulting functions with the sources of the defaults, ie, how did you work these out, is there anywhere in the code we can link to for the defaulting in the API so that we can see if any of these change easily at a later date.

Would also be good to get some test cases for these changes. This area is sparsely tested and without proper tests we aren't going to be able to ship this.

**Comment by Danil-Grigorev** [comment]
> > Another approach could be use generation comparison, and drop defaulting/spec hashing instead, but this way the operator will frequently poll API server for updates without value changes.
> 
> Does the operator use an informer? If such, this shouldn't be a problem, we can use a cached reader and that should be sufficient.
> 
> If we have informers set up already for this, then we should consider if the generation approach may be simpler, I believe this is what other implementation in library go have been using as one of their indicators, there must be a good reason for that. We may want to speak to the API folks to try and work out a bit more about the history of the apply code

Yes, the implementation is intended to be used with an informer. I'll see how far I could test current implementation, and I think it would be good to merge it this release with the approach we already choose. Then the upstream discussion could raise the question, and the `library-go` implementation could implement this in my opinion simpler approach (assuming we won't be able to figure out reasoning to do otherwhise). If you agree I'll open a JIRA ticket to that.

For now, I'll see how well tested the current implementation is, and do the changes. Do you think some heavier refactoring is allowed here? The `observedGeneration` is a dead code with the introduced changes, so it's either one way or another if we'd like to be clear.

**Comment by Danil-Grigorev** [comment]
> Would be good to see some comments on the defaulting functions with the sources of the defaults, ie, how did you work these out, is there anywhere in the code we can link to for the defaulting in the API so that we can see if any of these change easily at a later date.
> 
> Would also be good to get some test cases for these changes. This area is sparsely tested and without proper tests we aren't going to be able to ship this.

@JoelSpeed I added the comments on the code. The code is currently tested as much as it could be with the structure we got. Most of the absent coverage (38%) comes from `structured <-> unstructured` conversion, and error checking. I would adjust the code to ignore errors at these points, as we expect a structured resource on the methods like `apply`, so there should be no error-prone behavior there.

File: pkg/operator/webhook_configuration_helper.go [hunk 1]
```diff
@@ -152,3 +153,104 @@ func jsonPatchNoError(original, modified runtime.Object) string {
...
+		obj.Scope = &s
+	}
+}
+
+// setDefaults_ServiceReference sets defaults for Webhook's ServiceReference
+func setDefaultsServiceReference(obj *admissionregistrationv1.ServiceReference) {
+	if obj.Port == nil {
+		obj.Port = pointer.Int32Ptr(443)
+	}
+}
```
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply pkg/operator/webhook_configuration_helper.go:256]: @JoelSpeed Are we ok importing kubernetes into our repo?

File: pkg/operator/sync_test.go [hunk 2]
```diff
@@ -139,6 +140,7 @@ type webhookTestCase struct {
 
 func TestSyncValidatingWebhooks(t *testing.T) {
 	defaultConfiguration := mapiv1.NewValidatingWebhookConfiguration()
+	setDefaultsValidatingWebhookConfiguration(defaultConfiguration)
```
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply pkg/operator/sync_test.go:143]: It is interfering with testing. If I don't set defaults here then I'm providing a webhook configuration that is not fully populated, and as we are using a fake client here, nothing will fix it. So all update tests will independently on the changes I do or don't do, will execute the update operation. 

Another option would be putting this defaulting into `api` package, and apply during `NewValidatingWebhookConfiguration` call. But if we choose any other approach, we would have to make changes to our API, which is not good and under practices.

File: pkg/operator/webhook_configuration.go [hunk 3]
```diff
@@ -183,33 +196,27 @@ func applyUnstructured(resourceClient dynamic.ResourceInterface, path string, re
...
 	}
 
 	// at this point we know that we're going to perform a write.  We're just trying to get the object correct
 	toWrite := existingCopy // shallow copy so the code reads easier
-	if err := unstructured.SetNestedField(toWrite.Object, requiredSpec, path); err != nil {
-		return nil, false, err
-	}
+	// Ignore the error, as the object types are identical, and the path
+	// in the resource was already found
+	unstructured.SetNestedField(toWrite.Object, requiredSpec, path)
```
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply pkg/operator/webhook_configuration.go]: You are right, it presense was justified by the code which didn't belong there, so I removed it. Now we are returning the error as expected.

**Comment by JoelSpeed** [comment]
/retitle Reconciliation of MutatingWebhookConfiguration values didnt happen

This shouldn't be needed to resolve the attached bug, let's try to get it verified with just the other PR and see if that goes through

File: pkg/operator/webhook_configuration.go [hunk 4]
```diff
@@ -21,54 +22,58 @@ const genericCABundleInjectorLabel = "service.beta.openshift.io/inject-cabundle"
 // it will be created. If it does exist, the metadata of the required
 // mutatingwebhookconfiguration will be merged with the existing mutatingwebhookconfiguration
 // and an update performed if the mutatingwebhookconfiguration spec and metadata differ from
-// the previously required spec and metadata. For further detail, check the top-level comment.
-func applyMutatingWebhookConfiguration(client dynamic.Interface, recorder events.Recorder,
+// the previously required spec and metadata based on generation change.
+func applyMutatingWebhookConfiguration(client v1.MutatingWebhookConfigurationInterface, recorder events.Recorder,
```
Comment (JoelSpeed) [inline pkg/operator/webhook_configuration.go:26]: I would suggest just using `kubernetes.Interface` and then making into a `MutatingWebhookConfigurationInterface` within the code itself

File: pkg/operator/webhook_configuration.go [hunk 5]
```diff
@@ -182,39 +115,48 @@ func applyUnstructured(resourceClient dynamic.ResourceInterface, path string, re
...
+		reportUpdateEvent(recorder, required, err)
+		return actual, true, nil
+	}
+	return actual, false, nil
+}
 
-	if klog.V(4) {
-		klog.Infof("%s %q changes: %v", required.GetKind(), required.GetNamespace()+"/"+required.GetName(), jsonPatchNoError(existing, toWrite))
+func copyValidatingWebhookCABundle(existing, required *admissionregistrationv1.ValidatingWebhookConfiguration) {
+	existingMutatingWebhooks := make(map[string]admissionregistrationv1.ValidatingWebhook)
```
Comment (JoelSpeed) [inline pkg/operator/webhook_configuration.go:149]: This should probably be validating 😛 
```suggestion
	existingValidatingWebhooks := make(map[string]admissionregistrationv1.ValidatingWebhook)
```

Comment (JoelSpeed) [inline pkg/operator/webhook_configuration.go:149]: This should be validating
```suggestion
	existingValidatingWebhooks := make(map[string]admissionregistrationv1.ValidatingWebhook)
```

File: pkg/operator/suite_test.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/envtest/printer"
+)
+
+var (
+	cfg     *rest.Config
+	testEnv *envtest.Environment
+)
+
+func TestMachinesetController(t *testing.T) {
+	RegisterFailHandler(Fail)
```
Comment (JoelSpeed) [inline pkg/operator/suite_test.go:20]: If we're using klog at all within this code, let's set the output
```suggestion
	RegisterFailHandler(Fail)
	klog.SetOutput(GinkgoWriter)
```

**Review by JoelSpeed — COMMENTED** [review]
Could do with some testing ideally, see comments inline

File: pkg/apis/machine/v1beta1/machine_webhook.go [hunk 7]
```diff
@@ -128,8 +128,12 @@ const (
...
 	// webhookFailurePolicy is ignore so we don't want to block machine lifecycle on the webhook operational aspects.
 	// This would be particularly problematic for chicken egg issues when bootstrapping a cluster.
-	webhookFailurePolicy = admissionregistrationv1.Ignore
-	webhookSideEffects   = admissionregistrationv1.SideEffectClassNone
+	webhookFailurePolicy      = admissionregistrationv1.Ignore
+	webhookSideEffects        = admissionregistrationv1.SideEffectClassNone
+	webhookMatchPolicy        = admissionregistrationv1.Equivalent
+	webhookResourceSelector   = metav1.LabelSelector{}
+	webhookScope              = admissionregistrationv1.NamespacedScope
+	webhookReinvocationPolicy = admissionregistrationv1.IfNeededReinvocationPolicy
```
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply pkg/apis/machine/v1beta1/machine_webhook.go:135]: It is not a default one, but I set it explicitly as someone could try to negate our changes from mutation with his own webhook. This way our webhook will be invoked for each additional call.

**Comment by Danil-Grigorev** [comment]
/close

Closing this in favor of Joel's implementation currently merged https://github.com/openshift/machine-api-operator/pull/707 and an upcoming library-go approach - https://github.com/openshift/library-go/pull/836

**Pull request closed without merging**